### PR TITLE
fix: missing vault status for some primes

### PR DESCRIPTION
**Pull request by SlayerOrnstein**
### What did you fix? 
This fixes `vault` missing for some prime warframes. i.e. Nidus Prime was missing `vault` due to data not being available on ogg. Vault status is now retrieved from the wikia and fallsback to ogg.

---

### Reproduction steps

---

### Evidence/screenshot/link to line

image

### Considerations
- Does this contain a new dependency? **Yes**
- Does this introduce opinionated data formatting or manual data entry? **No?**
- Does this pr include updated data files in a separate commit that can be reverted for a clean code-only pr? **No**
- Have I run the linter? **Yes**
- Is is a bug fix, feature request, or enhancement? **Bug Fix**

### CODE REVIEW

File: build/parser.mjs [hunk 1]
```diff
@@ -941,12 +941,20 @@ class Parser {
...
    */
-  addVaultData(item, vaultData) {
+  addVaultData(item, category, wikiaData, vaultData) {
+    if (!['weapons', 'warframes', 'mods', 'upgrades'].includes(category.toLowerCase())) return;
     if (!item.name.endsWith('Prime')) return;
+
     const target = vaultData.find((i) => i.Name.toLowerCase() === item.name.toLowerCase());
 
+    if (item.type === 'Archwing') category = 'archwings';
+    if (category === 'Upgrades') category = 'mods';
```
Comment (AyAyEm) [inline build/parser.mjs:955]: reassignment of function parameters is not very good by design, in this case nothing bad should happen, but could you give another name to category in this context like vaultCategory? And also if you're changing function parameters to use in the rest of the function do it at the first chance you have, it'll be easier to read.

For example
```js
let vaultCategory = category;
if (item.type === 'Archwing') vaultCategory = 'archwings';
if (vaultCategory === 'Upgrades') vaultCategory = 'mods';

if (!['weapons', 'warframes', 'mods', 'upgrades'].includes(vaultCategory.toLowerCase())) return;
if (!item.name.endsWith('Prime')) return;

const target = vaultData.find((i) => i.Name.toLowerCase() === item.name.toLowerCase());
```

File: build/parser.mjs [hunk 2]
```diff
@@ -193,7 +193,7 @@ class Parser {
     this.addPatchlogs(result, data.patchlogs);
     this.addAdditionalWikiaData(result, category, data.wikia);
     this.detectPrime(result);
-    this.addVaultData(result, data.vaultData);
+    this.addVaultData(result, category, data.wikia, data.vaultData);
```
Comment (AyAyEm) [inline build/parser.mjs:196]: you forgot to update the line 223 with the new arguments

File: build/scraper.mjs [hunk 3]
```diff
@@ -1,17 +1,18 @@
...
 import ModScraper from './wikia/scrapers/ModScraper.mjs';
-import WeaponScraper from './wikia/scrapers/WeaponScraper.mjs';
-import WarframeScraper from './wikia/scrapers/WarframeScraper.mjs';
 import VersionScraper from './wikia/scrapers/VersionScraper.mjs';
-import readJson from './readJson.mjs';
+import WarframeScraper from './wikia/scrapers/WarframeScraper.mjs';
+import WeaponScraper from './wikia/scrapers/WeaponScraper.mjs';
 
 import { get, getJSON, retryAttempts } from './network.mjs';
+import ArchwingScraper from './wikia/scrapers/ArchwingScaper.mjs';
```
Comment (AyAyEm) [inline build/scraper.mjs:15]: could you revert these changes? It seems that your formatter thought it would be better to rearrenge them 😂

**Review by AyAyEm — CHANGES_REQUESTED** [review]
I only reviewed the code, i will do the review of the data if @TobiTenno approve the idea of adding more wikia dependency

File: build/parser.mjs [hunk 4]
```diff
@@ -968,9 +976,8 @@ class Parser {
     if (target.EstimatedVaultedDate) {
       item.estimatedVaultDate = this.normalizeOggDate(target.EstimatedVaultedDate);
     }
-    if (target.Vaulted) {
-      item.vaulted = target.Vaulted;
-    }
+
+    item.vaulted = wikiaItem.vaulted || target.vaulted;
```
Comment (AyAyEm) [inline build/parser.mjs:980]: is the target.vaulted that unreliable so we need to depend directly from the wikia if available?

File: build/parser.mjs (hunk 1 above)
Comment (SlayerOrnstein) [reply build/parser.mjs]: This is actually not allowed in dart either so not sure why I did it here 🤣

File: build/parser.mjs (hunk 2 above)
Comment (SlayerOrnstein) [reply build/parser.mjs:196]: oh

File: build/parser.mjs (hunk 4 above)
Comment (SlayerOrnstein) [reply build/parser.mjs:981]: if you want we could use ogg as the primary and the wikia second

Comment (AyAyEm) [reply build/parser.mjs:981]: nah i'm ok having the wikia first

Comment (TobiTenno) [reply build/parser.mjs:981]: i think i was saying it in a weird way that wasn't helpful. if ogg is defined, use ogg. if wiki is defined and ogg is not, use wiki... imo.

the fallback language, i think, was making my meaning unclear

File: build/scraper.mjs [hunk 5]
```diff
@@ -12,6 +12,7 @@ import VersionScraper from './wikia/scrapers/VersionScraper.mjs';
 import readJson from './readJson.mjs';
 
 import { get, getJSON, retryAttempts } from './network.mjs';
+import ArchwingScraper from './wikia/scrapers/ArchwingScaper.mjs';
```
Comment (TobiTenno) [inline build/scraper.mjs:15]: this should be above ModScraper... not sure why this got imported lie this

Comment (SlayerOrnstein) [reply build/scraper.mjs:15]: auto import ftw right

File: build/scraper.mjs [hunk 6]
```diff
@@ -134,9 +135,7 @@ class Scraper {
    * @property {string} textureLocation location of the image for the corresponding item
    */
   /**
-   * @typedef {Object} ImageManifest
-   * @property {ImageManifestItem[]} Manifest
-   */
+   * */
```
[the remark this answers is not in the document]
Comment (SlayerOrnstein) [reply build/scraper.mjs:138]: I'm not even sure what happened here. It looks like it was changed by the formatter but it removed that whole comment... I'm seeing why now and it says because it wasn't declared so I'm deferring to you if you want me to clean up the empty comment or add it back

File: build/scraper.mjs [hunk 7]
```diff
@@ -227,8 +229,7 @@ class Scraper {
 
   /**
    * Formatted date string. Format: "YYYY MM DD"
-   * @typedef {string} OggDateStamp
-   */
+   *    */
```
Comment (TobiTenno) [inline build/scraper.mjs:232]: wth is vs code doing. please stop.

**Review by TobiTenno — CHANGES_REQUESTED** [review]
seems fine overall, just some weird formatting on vs code

File: build/scraper.mjs (hunk 7 above)
Comment (SlayerOrnstein) [reply build/scraper.mjs:232]: it's actually off. Looks like this was after I ran lint:fix

File: build/parser.mjs (hunk 4 above)
Comment (SlayerOrnstein) [reply build/parser.mjs:981]: so no more

File: build/scraper.mjs (hunk 7 above)
Comment (TobiTenno) [reply build/scraper.mjs]: right, but lint doesn't normally fix what you've already got in the comment... i'd actually rather we not get rid of our typedef for OggDateStamp...

File: build/scraper.mjs (hunk 6 above)
[the remark this answers is not in the document]
Comment (TobiTenno) [reply build/scraper.mjs]: I'd prefer it stick around. typedefs exist to create type definitions that are for the internal mechanism of the library in this case for a bunch of stuff that exist for their own sake. i'm not sure why vs code is telling you it's not defined... this is how typedefs work in jsdoc, lol

File: build/scraper.mjs [hunk 8]
```diff
@@ -133,10 +134,6 @@ class Scraper {
    *    identifier corresponding to a {@link module:warframe-items.Item.uniqueName|uniqueName}
    * @property {string} textureLocation location of the image for the corresponding item
    */
-  /**
-   * @typedef {Object} ImageManifest
-   * @property {ImageManifestItem[]} Manifest
-   */
```
Comment (TobiTenno) [inline build/scraper.mjs:139]: this is an internal type def and should not be removed

File: build/scraper.mjs [hunk 9]
```diff
@@ -227,7 +227,6 @@ class Scraper {
 
   /**
    * Formatted date string. Format: "YYYY MM DD"
-   * @typedef {string} OggDateStamp
```
Comment (TobiTenno) [inline build/scraper.mjs:230]: this is an internal typedef and should not be removed

Comment (SlayerOrnstein) [reply build/scraper.mjs]: Just pushed out the commit that adds it back

File: build/parser.mjs [hunk 10]
```diff
@@ -941,11 +941,20 @@ class Parser {
    * Adds releaseDate, vaultDate and estimatedVaultDate to all primes using
    * data from "Ducats or Plat".
    * @param {Item} item data to append vault data to
+   * @param {module:warframe-items.Category} category of the data
+   * @param {WikiaData} wikiaData from wikia to apply
    * @param {VaultData} vaultData to look up data for the {@param item}
    */
-  addVaultData(item, vaultData) {
+  addVaultData(item, category, wikiaData, vaultData) {
```
Comment (TobiTenno) [inline build/parser.mjs:948]: i'm not sure why the category param is needed, we should already have the category defined from the item

also, generally you add new params at the end of the signature, not in the middle, unless you're adding required parameters before optional parameters.... but all of these are required

Comment (SlayerOrnstein) [reply build/parser.mjs:948]: It's not but because I don't understand javascript as much a fair bit was done from what I understood in `addAdditionalWikiaData`, I did want to originally use `item.category` or `item.type` but because I didn't see it being used I assumed there was a reason for it and just added it here, which is about the same explanation for the order actually

Comment (TobiTenno) [reply build/parser.mjs:948]: ok, what about the stuff that can be vaulted? you can't vault resources...

Comment (SlayerOrnstein) [reply build/parser.mjs:948]: skill issues on my part then. but yeah I got you

Comment (TobiTenno) [reply build/parser.mjs]: nono, this is definitely a cultural js thing that's developed over time that's not really well communicated

Comment (SlayerOrnstein) [reply build/parser.mjs]: still missing something because now there's no archwings in wikiadata

File: build/parser.mjs [hunk 11]
```diff
@@ -968,9 +978,8 @@ class Parser {
     if (target.EstimatedVaultedDate) {
       item.estimatedVaultDate = this.normalizeOggDate(target.EstimatedVaultedDate);
     }
-    if (target.Vaulted) {
-      item.vaulted = target.Vaulted;
-    }
+
+    item.vaulted = target.vaulted || wikiaItem.vaulted;
```
Comment (AyAyEm) [inline build/parser.mjs:982]: i've started reviewing the data and it seems like this line should merge objects if both are available, we're missing a bunch of 'wikiaThumbnail' entries and other data, also the parent data is all mangled

[itemsOld.json](https://github.com/WFCD/warframe-items/files/13850184/itemsOld.json)
[itemsNew.json](https://github.com/WFCD/warframe-items/files/13850183/itemsNew.json)
`curl https://termbin.com/7yl0`
`jsondiffpatch /itemsOld.json ./itemsNew.json`
if you guys know of a better solution to diff json let me know

Comment (SlayerOrnstein) [reply build/parser.mjs:982]: hmm I'm not entirely sure. Like I said not super versed in js but I don't really see this one would affect wikiaThumbnail (sentence feels a little off but I'm genuinely asking)

Comment (AyAyEm) [reply build/parser.mjs:982]: Sorry my mistake i forgot some entries just are kinda random when you build

Comment (AyAyEm) [reply build/parser.mjs:982]: and also idk why i thought this line was an object it's a boolean, but still some weird thing is happening and valkyr prime doesn't have the vaulted entry anymore

Comment (SlayerOrnstein) [reply build/parser.mjs:982]: That one I will check

**Comment by AyAyEm** [comment]
All the pertinent data for this PR
[itemsOld.json](https://github.com/WFCD/warframe-items/files/13850184/itemsOld.json)
[itemsNew.json](https://github.com/WFCD/warframe-items/files/13850183/itemsNew.json)

[vaultData.json](https://github.com/WFCD/warframe-items/files/13850515/vaultData.json)
[wikia.json](https://github.com/WFCD/warframe-items/files/13850516/wikia.json)

File: build/parser.mjs (hunk 11 above)
Comment (AyAyEm) [reply build/parser.mjs:982]: it's a casing issue i think
```suggestion
    item.vaulted = target.Vaulted || wikiaItem.vaulted;
```

Comment (AyAyEm) [reply build/parser.mjs:982]: i rebuild and it didn't fix it uhhh

Comment (AyAyEm) [reply build/parser.mjs:982]: oh it actually fixed it my mistaked

Comment (SlayerOrnstein) [reply build/parser.mjs]: oh well it should be `target.Vaulted` so yeah?

Comment (AyAyEm) [reply build/parser.mjs]: Yeah all the data from ogg seems to be in pascalcase

File: build/parser.mjs [hunk 12]
```diff
@@ -941,11 +941,21 @@ class Parser {
...
+   * @param {module:warframe-items.Category} category of the data
+   * @param {WikiaData} wikiaData from wikia to apply 
    */
-  addVaultData(item, vaultData) {
+  addVaultData(item, vaultData, category, wikiaData) {
+    let vaultCategory = category;
+    if (item.type === 'Archwing') vaultCategory = 'archwings';
+    if (category === 'Upgrades') vaultCategory = 'mods';
+
+    if (!['weapons', 'warframes', 'mods', 'upgrades', 'archwings'].includes(vaultCategory.toLowerCase())) return;
```
Comment (AyAyEm) [inline build/parser.mjs:954]: Sentinels also have primes
```suggestion
    if (!['weapons', 'warframes', 'mods', 'upgrades', 'archwings', 'sentinels'].includes(vaultCategory.toLowerCase())) return;
```

File: build/parser.mjs [hunk 13]
```diff
@@ -941,11 +941,21 @@ class Parser {
    * Adds releaseDate, vaultDate and estimatedVaultDate to all primes using
    * data from "Ducats or Plat".
    * @param {Item} item data to append vault data to
+
```
Comment (AyAyEm) [inline build/parser.mjs:944]: Unnecessary blank line

**Comment by AyAyEm** [comment]
is it possible to get the vaulted status from relics from wikia? if so would you be willing to do the work in this pr @SlayerOrnstein ? (you've already done alot this is totally optional) if so we could fix #496

**Comment by SlayerOrnstein** [comment]
Hmm it would be but I think that's a feature better suited for warframe-relics then add that into items

File: build/parser.mjs (hunk 12 above)
Comment (SlayerOrnstein) [reply build/parser.mjs]: Adding sentinels means adding a scraper for companions (https://warframe.fandom.com/wiki/Module:Companions/data). I can add this but I'm not sure if this would break anything else that would include companions.

Comment (SlayerOrnstein) [reply build/parser.mjs:953]: also looks like they also already have it https://api.warframestat.us/items/search/Carrier%20Prime/

Comment (AyAyEm) [reply build/parser.mjs:953]: Well this pr is removing the data that sentinels already have, so maybe just use the ogg for it and not the wikia? or also add the  wikia data?

File: build/parser.mjs [hunk 14]
```diff
@@ -968,9 +977,8 @@ class Parser {
     if (target.EstimatedVaultedDate) {
       item.estimatedVaultDate = this.normalizeOggDate(target.EstimatedVaultedDate);
     }
-    if (target.Vaulted) {
-      item.vaulted = target.Vaulted;
-    }
+
+    item.vaulted = target.Vaulted || wikiaItem.vaulted;
```
Comment (AyAyEm) [inline build/parser.mjs:981]: Just realized something, this shouldn't be a boolean or, because it's fine if it's false
```suggestion
    item.vaulted = typeof target.Vaulted === 'boolean' ? target.Vaulted : wikiaItem.vaulted;
```

File: build/parser.mjs (hunk 12 above)
Comment (AyAyEm) [reply build/parser.mjs:953]: Just diff the data that i provided and you'll see the problem
`jsondiffpatch ./itemsOld.json ./itemsNew.json`

File: build/parser.mjs [hunk 15]
```diff
@@ -942,10 +942,19 @@ class Parser {
...
+    if (item.type === 'Archwing') vaultCategory = 'archwings';
+    if (category === 'Upgrades') vaultCategory = 'mods';
+
+    if (!['weapons', 'warframes', 'mods', 'upgrades', 'archwings'].includes(vaultCategory.toLowerCase())) return;
     if (!item.name.endsWith('Prime')) return;
+
     const target = vaultData.find((i) => i.Name.toLowerCase() === item.name.toLowerCase());
+    const wikiaItem = wikiaData[vaultCategory.toLowerCase()].filter((i) => i).find((i) => i.name === item.name);
 
     if (!target) {
```
Comment (AyAyEm) [inline build/parser.mjs:959]: @TobiTenno should wikiaItem be added to this test? So it'll fallback to wikia data in case the ogg data doesn't exist?
```suggestion
    if (!target && !wikiaItem) {
```

Comment (SlayerOrnstein) [reply build/parser.mjs:965]: So wikia doesn't have `EstimatedVaultedDate` or `VaultedDate` so if it doesn't exist it would be better to just skip over it since there's no other way to add it from wikia?

File: build/parser.mjs (hunk 14 above)
Comment (SlayerOrnstein) [reply build/parser.mjs:987]: unless the ogg data is wrong or the wikiadata is wrong? that was I figured would happen in either case. if both are false then that's fine

File: build/parser.mjs (hunk 15 above)
Comment (AyAyEm) [reply build/parser.mjs:965]: Well if it's skipped it won't add the vaulted state right?

File: build/parser.mjs (hunk 14 above)
Comment (AyAyEm) [reply build/parser.mjs:987]: it's because in js undefined is cast to false in a boolean comparison, so if we have the target data and vault is false it'll ignore it and look for the wikia which could be false, true or undefined. If ogg is the default data source it doesn't make sense to ignore it when the vaulted state is false.

File: build/parser.mjs (hunk 15 above)
Comment (SlayerOrnstein) [reply build/parser.mjs:965]: Yup but it also won't error out when target doesn't exist but Wikia does

File: build/parser.mjs (hunk 14 above)
Comment (SlayerOrnstein) [reply build/parser.mjs:987]: Ah I understand now

Comment (AyAyEm) [reply build/parser.mjs:987]: basically we take ogg data by default unless it's not a boolean (undefined) so if it's not we take whatever the wikia gives us undefined, false, true, in the previous scenario we would take true from ogg and if it's not true we would take undefined, true or false from wikia.

File: build/parser.mjs (hunk 15 above)
Comment (AyAyEm) [reply build/parser.mjs:965]: i see you're right we could just add `target?.ReleaseDate` in that case so it won't error out, because i think being vaulted or not is the priority and not the dates.

Comment (SlayerOrnstein) [reply build/parser.mjs:965]: oooo I didn't know I could use that in JS, this is null check right?

Comment (AyAyEm) [reply build/parser.mjs:965]: yep it just ignores the next dot if it's undefined or null it's called Optional Chaining
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining

File: build/parser.mjs (hunk 14 above)
Comment (AyAyEm) [reply build/parser.mjs:987]: Just remembered there's also nullish coalescing operator
```suggestion
    item.vaulted = target.Vaulted ?? wikiaItem.vaulted;
```

File: build/parser.mjs (hunk 15 above)
Comment (TobiTenno) [reply build/parser.mjs:965]: yeah

File: build/parser.mjs (hunk 14 above)
Comment (TobiTenno) [reply build/parser.mjs:987]: this would be better, since `target.Vaulted` could be false, in which case you'd want to use that.

File: build/scraper.mjs (hunk 8 above)
Comment (TobiTenno) [reply build/scraper.mjs:139]: still waiting on this to come back

**Review by TobiTenno — CHANGES_REQUESTED** [review]
if  you're gonna add sentinels & archwings, you should upgrade this to a feat, or do those in a separate PR.

**Comment by SlayerOrnstein** [comment]
> if you're gonna add sentinels & archwings, you should upgrade this to a feat, or do those in a separate PR.

Gonna have to be a feature then because without archwings building breaks and without sentinels data is removed

File: build/scraper.mjs (hunk 8 above)
Comment (SlayerOrnstein) [reply build/scraper.mjs:139]: gimme a few lost my gpg keys

**Comment by TobiTenno** [comment]
> > if you're gonna add sentinels & archwings, you should upgrade this to a feat, or do those in a separate PR.
> 
> Gonna have to be a feature then because without archwings building breaks and without sentinels data is removed

aren't both of those only because you added those to being checked in the wiki updates?

those both seem like things that should be in their own features...

**Comment by SlayerOrnstein** [comment]
> > > if you're gonna add sentinels & archwings, you should upgrade this to a feat, or do those in a separate PR.
> > 
> > 
> > Gonna have to be a feature then because without archwings building breaks and without sentinels data is removed
> 
> aren't both of those only because you added those to being checked in the wiki updates?
> 
> those both seem like things that should be in their own features...

Because Warframes and Archwings are under the same category it would try to look for Odonata Prime under warframes. I can if you want have it ignore Archwings in this PR and then add it back in a different PR with sentinels but I guess because I still have to check the category before calling `wikiaItem` I would have to ignore sentinels and possible still end up removing vault fields from sentinels

**Comment by SlayerOrnstein** [comment]
So I'll make this one depend on #506

File: build/scraper.mjs (hunk 8 above)
Comment (TobiTenno) [reply build/scraper.mjs]: :fearful:

Comment (SlayerOrnstein) [reply build/scraper.mjs]: Line should be back now? Also I thought I did, forgot I put the private and public key in the yubi

File: build/parser.mjs [hunk 16]
```diff
@@ -959,18 +968,18 @@ class Parser {
...
-    if (target.EstimatedVaultedDate) {
+    if (target?.EstimatedVaultedDate) {
       item.estimatedVaultDate = this.normalizeOggDate(target.EstimatedVaultedDate);
     }
-    if (target.Vaulted) {
-      item.vaulted = target.Vaulted;
-    }
+
+    item.vaulted = target?.Vaulted ?? wikiaItem.vaulted;
+    if (item.vaulted === 'N/A') item.vaulted = true;
```
Comment (AyAyEm) [inline build/parser.mjs:982]: Could you explain this line? Like in which case it would be N/A and why?

Comment (AyAyEm) [reply build/parser.mjs:982]: i don't like much having this one line of code for one warframe but i guess there's no other way around it

Comment (TobiTenno) [reply build/parser.mjs:982]: theoretically would also match other stuff if they ever move Vaulted to a field on item

Comment (SlayerOrnstein) [reply build/parser.mjs:982]: I would imagine for primes that come from baro since technically they aren't vaulted but also not unvaulted either

**Comment by TobiTenno** [comment]
i see 3 things that should not be in the PR

**Review by TobiTenno — APPROVED** [review]

**Review by AyAyEm — APPROVED** [review]
Looks good to me
[vaulted.csv](https://github.com/WFCD/warframe-items/files/14132711/vaulted.csv)
